### PANDAS LIB (Python data analysis library; takes data & creates a Python object w/ rows & columns called 'data frame' == table in statistical software)

In [ ]:
#### Use case for pandas (instead of using traditional DB): 
#### 1. Simple 1-step solutions
#### 2. Easier to do visualizations - retaining formatting for user experience
#### 3. For large volume data - better to do this in spark

In [14]:
import pandas as pd #import pandas lib

pd.set_option('display.max_columns', 200) #sets display cols in a data frame; otherwise rows will overflow
pd.set_option('display.max_rows', 200) #sets display cols in a data frame; otherwise rows will overflow
surveyResponse = pd.read_csv('Survey_Resp.csv') #surveyResponse is the primary data frame in this notebook
surveyResponse.head(10)

RESP_ID  SUR_SCORE         MASKED_IP          CRT_DT COUNTRY_CODE  \
0  13071656          8  MTQ5,Dk1,TMy,A==  1/18/2018 8:48           FI   
1  13071811         10  MTM0Mjk4,TE0OA==  1/18/2018 8:59           FR   
2  13071889          9      MzY4,DgzOTU=  1/18/2018 9:05           AE   
3  13071909         10  MTM4MzU5MjY2Mg==  1/18/2018 9:06           CZ   
4  13071920         10  MzU3ODcz,Tg1OA==  1/18/2018 9:07           IE   
5  13071988         10  Mjk1,jk1ODE4OQ==  1/18/2018 9:11           HU   
6  13072012          8  MTM2ODUxMTE4,g==  1/18/2018 9:12           GB   
7  13072097          7  MzA1ODE0,zYwOQ==  1/18/2018 9:17           IN   
8  13072253         10  MTg0,DEwMzMx,A==  1/18/2018 9:27           GB   
9  13072647         10  MTI5,zcxMTI0Mg==  1/18/2018 9:50           GB   

          REGION  
0        uusimaa  
1  ile-de-france  
2          dubai  
3          praha  
4            NaN  
5       budapest  
6            NaN  
7  uttar pradesh  
8            NaN  
9            NaN

In [15]:
surveyResponse.columns

Index(['RESP_ID', 'SUR_SCORE', 'MASKED_IP', 'CRT_DT', 'COUNTRY_CODE',
       'REGION'],
      dtype='object')

In [17]:
headers = ["RESPONSE_ID", "SURVEY_SCORE", "MASKED_IP", "SURVEY_DATE", "COUNTRY_CODE", "REGION"]
surveyResponse.columns = headers
surveyResponse.dtypes

RESPONSE_ID      int64
SURVEY_SCORE     int64
MASKED_IP       object
SURVEY_DATE     object
COUNTRY_CODE    object
REGION          object
dtype: object

In [18]:
surveyResponse.head(10)

RESPONSE_ID  SURVEY_SCORE         MASKED_IP     SURVEY_DATE COUNTRY_CODE  \
0     13071656             8  MTQ5,Dk1,TMy,A==  1/18/2018 8:48           FI   
1     13071811            10  MTM0Mjk4,TE0OA==  1/18/2018 8:59           FR   
2     13071889             9      MzY4,DgzOTU=  1/18/2018 9:05           AE   
3     13071909            10  MTM4MzU5MjY2Mg==  1/18/2018 9:06           CZ   
4     13071920            10  MzU3ODcz,Tg1OA==  1/18/2018 9:07           IE   
5     13071988            10  Mjk1,jk1ODE4OQ==  1/18/2018 9:11           HU   
6     13072012             8  MTM2ODUxMTE4,g==  1/18/2018 9:12           GB   
7     13072097             7  MzA1ODE0,zYwOQ==  1/18/2018 9:17           IN   
8     13072253            10  MTg0,DEwMzMx,A==  1/18/2018 9:27           GB   
9     13072647            10  MTI5,zcxMTI0Mg==  1/18/2018 9:50           GB   

          REGION  
0        uusimaa  
1  ile-de-france  
2          dubai  
3          praha  
4            NaN  
5       budapest  
6            NaN  
7  uttar pradesh  
8            NaN  
9            NaN

In [23]:
surveyMissingVal = surveyResponse.isnull() #creating a new data frame based on isnull method acting on original data frame
surveyMissingVal

RESPONSE_ID  SURVEY_SCORE  MASKED_IP  SURVEY_DATE  COUNTRY_CODE  REGION
0           False         False      False        False         False   False
1           False         False      False        False         False   False
2           False         False      False        False         False   False
3           False         False      False        False         False   False
4           False         False      False        False         False    True
...           ...           ...        ...          ...           ...     ...
4783        False         False      False        False         False   False
4784        False         False      False        False         False   False
4785        False         False      False        False         False   False
4786        False         False      False        False          True    True
4787        False         False      False        False         False   False

[4788 rows x 6 columns]

In [24]:
for col in surveyMissingVal.columns.values.tolist():
    print("Column name: ", col)

Column name:  RESPONSE_ID
Column name:  SURVEY_SCORE
Column name:  MASKED_IP
Column name:  SURVEY_DATE
Column name:  COUNTRY_CODE
Column name:  REGION


In [26]:
#surveyMissingVal["SURVEY_SCORE"].value_counts()
surveyMissingVal["COUNTRY_CODE"].value_counts() #returns counts for distinct values in the column - similar to GROUP BY in sql

False    3994
True      794
Name: COUNTRY_CODE, dtype: int64

In [27]:
surveyMissingVal["REGION"].value_counts()

False    3215
True     1573
Name: REGION, dtype: int64

In [28]:
for col_name in surveyMissingVal.columns.values.tolist():
    print("Column name: ", col_name)
    print(surveyMissingVal[col_name].value_counts())
    print('--------------------------------------')

Column name:  RESPONSE_ID
False    4788
Name: RESPONSE_ID, dtype: int64
--------------------------------------
Column name:  SURVEY_SCORE
False    4788
Name: SURVEY_SCORE, dtype: int64
--------------------------------------
Column name:  MASKED_IP
False    4788
Name: MASKED_IP, dtype: int64
--------------------------------------
Column name:  SURVEY_DATE
False    4788
Name: SURVEY_DATE, dtype: int64
--------------------------------------
Column name:  COUNTRY_CODE
False    3994
True      794
Name: COUNTRY_CODE, dtype: int64
--------------------------------------
Column name:  REGION
False    3215
True     1573
Name: REGION, dtype: int64
--------------------------------------


#### Survey response data set: missing data findings
##### Based on the summary above, of 4788 rows of data - 2 columns have missing data:
<li> 1. "COUNTRY_CODE": 794 missing</li>
<li> 2. "REGION": 1573 missing </li>

#### EXPLORATORY DATA ANALYSIS

In [30]:
#Get descriptive stats in data set based on columns of interest
#Describe method from pandas lib to do this
surveyResponse[['SURVEY_SCORE', 'SURVEY_DATE', 'COUNTRY_CODE', 'REGION']].describe(include='all') 
print(type(surveyResponse))

<class 'pandas.core.frame.DataFrame'>


In [37]:
surveyResponse.head(10).sort_values(by=['SURVEY_SCORE'], ascending=False, inplace=True) 
#head() won't sort in data frame, only for screen
#inplace actually sorts the data in the data frame

RESPONSE_ID  SURVEY_SCORE         MASKED_IP     SURVEY_DATE COUNTRY_CODE  \
1     13071811            10  MTM0Mjk4,TE0OA==  1/18/2018 8:59           FR   
3     13071909            10  MTM4MzU5MjY2Mg==  1/18/2018 9:06           CZ   
4     13071920            10  MzU3ODcz,Tg1OA==  1/18/2018 9:07           IE   
5     13071988            10  Mjk1,jk1ODE4OQ==  1/18/2018 9:11           HU   
8     13072253            10  MTg0,DEwMzMx,A==  1/18/2018 9:27           GB   
9     13072647            10  MTI5,zcxMTI0Mg==  1/18/2018 9:50           GB   
2     13071889             9      MzY4,DgzOTU=  1/18/2018 9:05           AE   
0     13071656             8  MTQ5,Dk1,TMy,A==  1/18/2018 8:48           FI   
6     13072012             8  MTM2ODUxMTE4,g==  1/18/2018 9:12           GB   
7     13072097             7  MzA1ODE0,zYwOQ==  1/18/2018 9:17           IN   

          REGION  
1  ile-de-france  
3          praha  
4            NaN  
5       budapest  
8            NaN  
9            NaN  
2          dubai  
0        uusimaa  
6            NaN  
7  uttar pradesh

In [38]:
surveyResponse.head(10).sort_values(by=['SURVEY_SCORE'], ascending=True, na_position='first') #null position

RESPONSE_ID  SURVEY_SCORE         MASKED_IP     SURVEY_DATE COUNTRY_CODE  \
7     13072097             7  MzA1ODE0,zYwOQ==  1/18/2018 9:17           IN   
0     13071656             8  MTQ5,Dk1,TMy,A==  1/18/2018 8:48           FI   
6     13072012             8  MTM2ODUxMTE4,g==  1/18/2018 9:12           GB   
2     13071889             9      MzY4,DgzOTU=  1/18/2018 9:05           AE   
1     13071811            10  MTM0Mjk4,TE0OA==  1/18/2018 8:59           FR   
3     13071909            10  MTM4MzU5MjY2Mg==  1/18/2018 9:06           CZ   
4     13071920            10  MzU3ODcz,Tg1OA==  1/18/2018 9:07           IE   
5     13071988            10  Mjk1,jk1ODE4OQ==  1/18/2018 9:11           HU   
8     13072253            10  MTg0,DEwMzMx,A==  1/18/2018 9:27           GB   
9     13072647            10  MTI5,zcxMTI0Mg==  1/18/2018 9:50           GB   

          REGION  
7  uttar pradesh  
0        uusimaa  
6            NaN  
2          dubai  
1  ile-de-france  
3          praha  
4            NaN  
5       budapest  
8            NaN  
9            NaN